syntax = "proto3";

package grpc_server;

//import "google/protobuf/struct.proto";

service GameController {
    rpc Create(CreateGameRequest) returns (GameResponse) {}
    rpc PartialUpdate(CreateGamePartialUpdateRequest) returns (GameResponse) {}
}

service MoveController {
    rpc Create(CreateMoveRequest) returns (MoveResponse) {}
}

message CreateGameRequestUserInfo {
    int32 id = 1;
    string username = 2;
    string email = 3;
}

message CreateGamePartialUpdateRequest {
    int32 id = 1;
    string winner = 2;
    double user_1_points = 3;
    double user_2_points = 4;
    string status = 5;
}

message CreateGameRequest {
    int32 user_1 = 1;
    int32 user_2 = 2;
}

message GameResponse {
    int32 id = 1;
    CreateGameRequestUserInfo user_1_info = 2;
    CreateGameRequestUserInfo user_2_info = 3;
    bool user_1_turn = 4;
    string winner = 5;
    optional double user_1_points = 6;
    optional double user_2_points = 7;
    string start_at = 8;
    string finish_at = 9;
    string status = 10;
}

message CreateMoveRequest {
    int32 game = 1;
    int32 user = 2;
    int32 checker_id = 3;
    repeated string new_positions = 4;
    bool is_king = 5;
    bool is_white = 6;
    bool is_dead = 7;
    repeated int32 killed = 8;
}

message MoveResponse {
    int32 id = 1; 
    int32 game = 2;
    int32 user = 3;
    int32 checker_id = 4;
    repeated string new_positions = 5;
    bool is_king = 6;
    bool is_white = 7;
    bool is_dead = 8;
    bool is_last_move = 9;
}